backend/src/federation/index.ts

```diff
@@ -0,0 +1,122 @@
+
+    server.on('connection', function (socket: any) {
+      // noiseSocket is E2E between you and the other peer
+      // pipe it somewhere like any duplex stream
+      console.log('Remote public key', socket.remotePublicKey.toString('hex'))
+      // console.log("Local public key", noiseSocket.publicKey.toString("hex")); // same as keyPair.publicKey
+
+      socket.on('data', (data: Buffer) => console.log('data:', data.toString('ascii')))
+
+      // process.stdin.pipe(noiseSocket).pipe(process.stdout);
+    })
+
+    await server.listen()
+
+    setInterval(async () => {
+      console.log('Announcing on topic:', TOPIC.toString('hex'))
+      await node.announce(TOPIC, keyPair).finished()
+    }, ANNOUNCETIME)
+
+    let successfulRequests: string[] = []
+    let errorfulRequests: string[] = []
+
+    setInterval(async () => {
+      console.log('Refreshing successful nodes')
+      successfulRequests = []
+    }, SUCCESSTIME)
+
+    setInterval(async () => {
+      console.log('Refreshing errorful nodes')
+      errorfulRequests = []
+    }, ERRORTIME)
+
+    setInterval(async () => {
+      const result = await node.lookup(TOPIC)
+
+      const collectedPubKeys: string[] = []
+
+      for await (const data of result) {
+        /* console.log(
+           `${data.from.host}:${data.from.port}: peers: ${data.peers.length}`
+           ); */
+        data.peers.forEach((peer: any) => {
+          const pubKey = peer.publicKey.toString('hex')
+          if (
+            pubKey !== keyPair.publicKey.toString('hex') &&
+            !successfulRequests.includes(pubKey) &&
+            !errorfulRequests.includes(pubKey) &&
+            !collectedPubKeys.includes(pubKey)
+          ) {
+            collectedPubKeys.push(peer.publicKey.toString('hex'))
+          }
+        })
+      }
+
+      console.log('Found new peers: ', collectedPubKeys)
+
+      collectedPubKeys.forEach((remotePubKey) => {
+        // publicKey here is keyPair.publicKey from above
+        const socket = node.connect(Buffer.from(remotePubKey, 'hex'))
+
+        /* socket.once("connect", function () {
+           console.log("client side emitted connect");
+           }); */
+
+        /* socket.once("end", function () {
+           console.log("client side ended");
+           }); */
+
+        socket.once('error', (err: any) => {
+          errorfulRequests.push(remotePubKey)
+          console.log(`error on peer ${remotePubKey}: ${err.message}`)
+        })
+
+        socket.on('open', function () {
+          // noiseSocket fully open with the other peer
+          // console.log("writing to socket");
+          socket.write(Buffer.from(`${nodeRand}`))
+          socket.write(Buffer.from(JSON.stringify(nodeAPI)))
+          successfulRequests.push(remotePubKey)
+        })
+        // pipe it somewhere like any duplex stream
+        // process.stdin.pipe(noiseSocket).pipe(process.stdout)
+      })
+    }, POLLTIME)
+  } catch (err) {
+    console.log(err)
+  }
+}
```

Comment:
it seems to be clear logically, but its hard to give a statement about correctness or functionality with a unknown framework. So we will see and research by using it.